File: computing/interview/design-patterns/creational/factory-method/csharp/FactoryMethodDemo.cs
using System;

namespace Interview.DesignPatterns.Creational.FactoryMethod
{
    public abstract class Report
    {
        public abstract void Export();
    }

    public sealed class PdfReport : Report
    {
        public override void Export() => Console.WriteLine("Exporting PDF report");
    }

    public sealed class CsvReport : Report
    {
        public override void Export() => Console.WriteLine("Exporting CSV report");
    }

    public abstract class ReportCreator
    {
        public void BuildReport()
        {
            var report = CreateReport();
            report.Export();
        }

        protected abstract Report CreateReport();
    }

    public sealed class PdfReportCreator : ReportCreator
    {
        protected override Report CreateReport() => new PdfReport();
    }

    public sealed class CsvReportCreator : ReportCreator
    {
        protected override Report CreateReport() => new CsvReport();
    }

    public static class Demo
    {
        public static void Main()
        {
            ReportCreator pdfCreator = new PdfReportCreator();
            ReportCreator csvCreator = new CsvReportCreator();

            pdfCreator.BuildReport();
            csvCreator.BuildReport();
        }
    }
}

File: computing/interview/design-patterns/creational/factory-method/java/FactoryMethodDemo.java
package interview.designpatterns.creational.factorymethod;

abstract class Report {
    abstract void export();
}

final class PdfReport extends Report {
    @Override
    void export() {
        System.out.println("Exporting PDF report");
    }
}

final class CsvReport extends Report {
    @Override
    void export() {
        System.out.println("Exporting CSV report");
    }
}

abstract class ReportCreator {
    public void buildReport() {
        Report report = createReport();
        report.export();
    }

    protected abstract Report createReport();
}

final class PdfReportCreator extends ReportCreator {
    @Override
    protected Report createReport() {
        return new PdfReport();
    }
}

final class CsvReportCreator extends ReportCreator {
    @Override
    protected Report createReport() {
        return new CsvReport();
    }
}

public final class FactoryMethodDemo {
    private FactoryMethodDemo() {}

    public static void main(String[] args) {
        ReportCreator pdf = new PdfReportCreator();
        ReportCreator csv = new CsvReportCreator();
        pdf.buildReport();
        csv.buildReport();
    }
}
